### PR TITLE
Ensure rank overlay plot starts at 0 even if not all bins present

### LINKED ISSUE #331: mcmc_rank_overlay: d_bin_counts doesn't contain rows for cases with n = 0
See also: https://discourse.mc-stan.org/t/rank-plots-dont-save-you-from-symmetric-non-stationary-chains/37624

In my case the first 4 bins all have count of 100 for the second chain, so the first chain's rank plot begins at 501. I suppose this could be a design choice but I think there should be bin entries in case n= 0.

image

### PR DESCRIPTION
**Pull request by jgabry**
fixes #331

Code to test: 

```r
x <- posterior::as_draws_array(
  list(
    list(theta = -2 + 0.003 * 1:1000 + arima.sim(list(ar = 0.7), n = 1000, sd = 0.5)),
    list(theta = 1 + -0.01 * 1:1000 + arima.sim(list(ar = 0.7), n = 1000, sd = 0.5))
  )
)
bayesplot::mcmc_rank_overlay(x)
```

### CODE REVIEW

File: R/mcmc-traces.R [hunk 1]
```diff
@@ -304,17 +304,24 @@ mcmc_rank_overlay <- function(x,
...
     count(.data$parameter, .data$chain, .data$bin_start)
 
+  # Now ensure that all combinations of parameter, chain, and bin_start exist
+  # even if no counts are present (https://github.com/stan-dev/bayesplot/issues/331)
+  all_params_chains <- dplyr::distinct(data, .data$parameter, .data$chain)
+  all_bins <- dplyr::distinct(histobins, .data$bin_start, .data$cut)
+  combos <- dplyr::cross_join(all_params_chains, all_bins)
+  d_bin_counts <- full_join(combos, d_bin_counts, by = c("parameter", "chain", "bin_start")) %>%
+    mutate(n = dplyr::coalesce(n, 0L))
+
```
Comment (jgabry) [inline R/mcmc-traces.R]: There may be a more elegant solution, but this seems to get the job done

**Comment by jgabry** [comment]
The failure on macOS-latest (devel) is due to the vdiffr package failing to compile on GitHub actions

**Comment by sims1253** [comment]
Maybe add a `set.seed` to the test to make it deterministic?

File: R/mcmc-traces.R (hunk 1 above)
Comment (sims1253) [reply R/mcmc-traces.R]: Not sure if this is more elegant but an expand.grid with one left join and an if_else to fill the zeros should also work. Something like the following. Reads a little easier for me, though that might just be me. (not tested!)
  ``` 
  all_combos <- tibble::as_tibble(expand.grid(
    parameter = unique(data$parameter),
    chain = unique(data$chain),
    bin_start = unique(histobins$bin_start),
    stringsAsFactors = FALSE
  ))

  d_bin_counts <- all_combos %>%
    left_join(d_bin_counts, by = c("parameter", "chain", "bin_start")) %>%
    mutate(n = if_else(is.na(n), 0L, n))
  ```

Comment (sims1253) [reply R/mcmc-traces.R]:
```
# Ensure all combinations exist and count values
d_bin_counts <- data %>%
  left_join(histobins, by = "value_rank") %>%
  count(.data$parameter, .data$chain, .data$bin_start) %>%
  right_join(
    dplyr::expand(.data$parameter, .data$chain, unique(histobins$bin_start)),
    by = c("parameter", "chain", "bin_start")
  ) %>%
  mutate(n = dplyr::coalesce(n, 0L))
  ```
  Or courtesy of claude, telling me my solution was verbouse and not dplyr-ish enough.

**Comment by jgabry** [comment]
> Maybe add a `set.seed` to the test to make it deterministic?

There's a `set.seed` in the file where I added the code to create the data for the test (it just doesn't show up in the diff since I didn't change it): 

https://github.com/stan-dev/bayesplot/blob/414755546305ea2ef06f70145c78712a024e8963/tests/testthat/data-for-mcmc-tests.R#L33

File: R/mcmc-traces.R (hunk 1 above)
Comment (jgabry) [reply R/mcmc-traces.R]: Thanks to you and Claude, both!

Comment (jgabry) [reply R/mcmc-traces.R]: > dplyr::expand

Except Claude didn't realize there is no `dplyr::expand` . There is `tidyr::expand`, but we don't already depend on `tidyr` so probably best not to add another dependency just for this.

**Comment by jgabry** [comment]
@sims1253 Thanks. I just pushed a commit that uses basically your first suggestion of `expand_grid` plus a `left_join`.

File: R/mcmc-traces.R (hunk 1 above)
Comment (sims1253) [reply R/mcmc-traces.R]: Ha, great help :D

**Pull request merged**